---
title: "BRIC vignette"
subtitle: BRIC (Biclustering-based gene Regulation Inference and Cell-type prediction)
abstract: BRIC is a novel biclustering method for the detection of the       repertoire of active GRS within each single cell. This introductory vignette provides an overview of the steps of the workflow, installation, and source of the dataset used as example.
date: "October 10, 2018"
output:
  BiocStyle::html_document:
    number_sections: no
    toc: yes
    highlight: pygments
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Intorduction to BRIC

BRIC (Biclustering-based gene Regulation Inference and Cell-type prediction) is a novel biclustering method for the detection of the repertoire of active GRS within each single cell, based on which we annotate the type and/or physiological state of each cell.

BRIC consists of two major steps: (i) detecting all gene co-regulation modules (GCM) each corresponding to a GRS, and (ii) inferring the active GRS of each cell, and further its cell type or physiological state. Step (i) first utilized our in-house left truncated mixture Gaussian (LTMG) model for an accurate quantification of the sample specific GRS for each individual gene. Step (ii) utilized the set of active states of GRS within a cell for its cell type prediction.  

# Requirements
## Environment

We will assum you have the following installed:

* A C++ 11 compatible compiler such as >=g++4.8(might work in g++-4.7, though untested)

* **make** which is also installed on most machines

## Input

The input to BRIC is the single-cell RNA-seq expression matrix:

* Rows correspond to genes and columns correspond to sampels(cells).
* Expression units: the preferred expression values are RPKM/FPKM/CPM. 
* The data file should be tab delimited.

# Installation

BRIC is based on 

1. our in-house biclustering algorithm for GRS specific gene modules detection

2. two R scripts for cell type inferring

Therefore, you will need to first install our biclustering algorithm, which is available at <https://github.com/maqin2001/BRIC/blob/master/BRIC_biclustering.zip>. Download the zip file , put it in any directpry, and type
```{r,engine='bash',eval=FALSE,unzip}
unzip BRIC_biclustering.zip
```
Go to the 'BRIC_biclustering' folder
```{r,engine='bash',eval=FALSE, cd}
cd BRIC_biclustering

```
Type `make` to compile the source code:
```{r,engine='bash',eval=FALSE, make}
make
```
Then the compiled codes are within the `BRIC_biclusering` directory.

# Example dataset

This tutorial run on a real dataset to illustrate the results obtained at each step.

As example, we will use a dataset containing 90 cells and 20,214 genes from human embryo.

> Yan, L. et al. Single-cell RNA-Seq profiling of human preimplantation embryos and embryonic stem cells. Nat. Struct. Mol. Biol. 20, 1131-1139 (2013)

The original expression matrix was downloaded from <https://s3.amazonaws.com/scrnaseq-public-datasets/manual-data/yan/nsmb.2660-S2.csv>. The expression is provided as RPKM value. For convenience, we removed the space in the column names and deleted the second column(Transcript_ID). The processed data is available at <https://github.com/maqin2001/BRIC/blob/master/Example%20data/Yan_expression.txt>. It is also included in the `data`subfolder of `BRIC_biclusering`.

The cell types are easy to tell from the cell names. For simplicity, we prepared a seperate file, which is available at <https://github.com/maqin2001/BRIC/blob/master/Example%20data/Yan_cell.csv> .

# Biclustering

The first step of BRIC workflow is to predict GRS specific GCMs. To do this we use our in-house biclustering tool. The input of this tool is gene expression matrix.

After download and compile the code as introduced in the **Installation** part, go to the `BRIC_biclusering` directory. You can either type the following command to run discretization + biclustering directly:
```{r,engine='bash',eval=FALSE, biclustering}
./qubic -i ./data/Yan_RPKM -R -f 0.85 -k 13 -c 0.90 -o 5000

```
This will result several output files, and among them you will find one named `Yan_RPKM.chars.blocks`,which will be used in the following step.

Or,you can first type the following command to discretize the data:
```{r,engine='bash',eval=FALSE,discretize}
./qubic -i ./data/Yan_RPKM -R -F

```
And then type the following command to run biclustering:
```{r,engine='bash',eval=FALSE, biclustering2}
./qubic -i ./data/Yan_RPKM.chars -d -f 0.85 -k 13 -c 0.90 -o 5000

```
This will also result a file named `Yan_RPKM.chars.blocks`. Note that the second way run discretization and biclustering seperately. Since discretization ususally take a long time, we recommend to use the second way. In this case, you can try biclustering parameters as needed based on the discretized data, avoiding wasting time on discretization.

(Here we use the example Yan_RPKM data which is located under the `data` folder. In the real case you can put your data at any folder you like, just specify the path to your data in the command)

# Construct weighted graph

The second step is to construct weighted graph based on the **.blocks** file generated from the biclustering step. You can use the `constructGraph.R` under the <https://github.com/maqin2001/BRIC/tree/master/cell_type_prediction>.
This will give you a file named **Yan_RPKM_graph.csv**. It should contain three columns, with the first two columns being node names and the third column denoting the weight between two corresponding nodes.
```{r construct graph,echo=FALSE }
setwd('C:/Users/Juan.XIe/Desktop/test1/')
files <-list.files(path=getwd(),pattern='*blocks',full.names=T, recursive=FALSE)

RAW <-read.table('Yan_RPKM',header=T,sep='\t')
F <-readLines(paste(getwd(),'/Yan_RPKM.chars.blocks',sep=''))  # read .blocks file

TEMP <-grep('Conds',F,value=T) ## extract condition lines in each BC
BC <-sapply(strsplit(TEMP,':',2),'[',2) # only keep cell names
temp1 <-gsub('#','_',BC)  # replace # with _
temp2 <-gsub('-','_',temp1)  
temp3 <-gsub('c','C',temp2) 
BC <-temp3

	CONDS <-as.character()   # store the conditions 
	label_C <-as.numeric()   # store the occurence of one condistions

	for (j in 1:length(BC)){
		BCcond <-unlist(strsplit(BC[j], split = " "))
		BCcond <-BCcond[BCcond!=""]  # exclude the blank string
		CONDS <-c(BCcond,CONDS)
		label_C <-c(label_C,rep(j,length(BCcond)))
	}

	df_C <-data.frame(conds=CONDS,label=label_C)
	uniq_C <-df_C$conds[!duplicated(df_C$conds)]   # unique conditions
	Node <-t(combn(uniq_C,2))
	
	Wt <-rep(-1,dim(Node)[1])
	for (k in 1:dim(Node)[1]){
		member1 <-df_C[which(df_C$conds %in% Node[k,1]),]   # identify which BC the k th Node appear
		member2 <-df_C[which(df_C$conds %in% Node[k,2]),]
		Wt[k] <-length(intersect(member1[,2],member2[,2])) # the weight between two node
	}
	GRAPH <-data.frame(Node1=Node[,1],Node2=Node[,2],weight=Wt)
	if (dim(GRAPH)[1]!=0)	{
	write.csv(subset(GRAPH,Wt!=0),"Yan_RPKM_graph.csv",row.names=FALSE)
	}

```
```{r show first lines of graph,echo=FALSE }
head(GRAPH)
```

# Clustering on the weighted graph

The last step is to do clustering on the weighted graph. Here we provide two commonly used clustering methods: MCL and spectral clustering. (you may use other clustering method), and the R script is available at <https://github.com/maqin2001/BRIC/blob/master/cell_type_prediction/clustering.R>. To run this script, you need the `igraph` , `MCL` and 'anocva' R package. You will also need `mclust` and `clues` R package if you have a reference label and want to calculate ARI.

Specifically, we will first read the weighted graph and convert it into adjacency matrix. Then we conduct either MCL or spectral clustering to cluster the cells.
```{r adj,message=FALSE}
setwd("C:/Users/Juan.XIe/Desktop/test1/") # you may change to your own folder
library(igraph)
library(mclust)
library(MCL)
library(clues)
library(anocva)

RAW <-read.table("Yan_RPKM",header=T,sep="\t")   # expression data
CellNum <-dim(RAW)[2]-1  # the number of cells 

Graph <-read.csv('Yan_RPKM_graph.csv',header=T,sep=",")
names(Graph) <-c('Node1','Node2','weight')
G <-graph.data.frame(Graph,directed = FALSE)  # convert edgelist into graph
A <- as_adjacency_matrix(G,type="both",attr="weight",names=TRUE,sparse=FALSE)  # convert graph into adjacency matrix
V_name <-rownames(A)   # the vertix
Covered <-length(V_name)  # the #of covered cells
```

```{r MCL,echo=FALSE,message=FALSE}
## MCL clustering on adjacency matrix
MCL_cs <-function(A){
	CLUST <-list()
	for (i in 1:100){
		CLUST[[i]] <-mcl(A,addLoops = FALSE,inflation =i,max.iter=200)
	}
	KK <- as.data.frame(do.call(rbind,lapply(CLUST,'[[',1)))  # extract the number of clusters
	CAN_I <-c(which(as.numeric(as.character(KK$V1))>=2)) 	# results that has more than 5 clusters
	tt <-as.numeric(as.character(KK$V1))
	tt <-sort(table(tt),decreasing=T)[1]
	Final_K <-as.numeric(names(tt))
		
	if (length(CAN_I)!=0){
		MATRIX <-rep(0,Covered)%o%rep(0,Covered)
		for (k in 1:length(CAN_I)){	
			MCL_label <-CLUST[[CAN_I[k]]]$Cluster  # record the label
			ClusterNum <-unique(MCL_label)   # record the number of clusters
			TEMP <-rep(0,Covered)%o%rep(0,Covered)
			temp <-rep(0,Covered) %o% rep(0,length(ClusterNum))
			for (n in 1:length(ClusterNum)){
				index <-which(MCL_label==ClusterNum[n])
				temp[index,n] <-1
				TEMP <-TEMP+temp[,n]%o%temp[,n] 
			}
		MATRIX <-MATRIX+TEMP
		}
		MATRIX <-MATRIX/length(CAN_I)
		rownames(MATRIX) <-colnames(MATRIX) <-rownames(A)
		hc <-hclust(dist(MATRIX))
		memb <-cutree(hc,k=Final_K)
		if (length(rownames(A)) ==CellNum){
			label <-memb
		}else{
			LEFT <-setdiff(names(RAW)[-1],V_name)
			LEFT_Cluster <-rep(Final_K+1,length(LEFT))
			df_cell_label <-data.frame(cell=c(names(memb),LEFT),cluster=c(memb,LEFT_Cluster),K=rep(Final_K+1,CellNum))				
			label <-df_cell_label$cluster
		}	
	}
return(label)
}

```


```{r spectral,echo=FALSE,message=FALSE}
## Spectral clustering on adjacency matrix, need provide the number of clusters
SC_cs <-function(A,K){   
	sc <-spectralClustering(A,k=K)
	names(sc) <-rownames(A)
	if (length(rownames(A)) ==CellNum){
		label <-sc
	}else{
		LEFT <-setdiff(names(RAW)[-1],V_name)
		LEFT_Cluster <-rep(K+1,length(LEFT))
		df_cell_label <-data.frame(cell=c(names(sc),LEFT),cluster=c(sc,LEFT_Cluster),K=rep(K+1,CellNum))				
		label <-df_cell_label$cluster
	}
return(label)
}

## users need to specify which clustering method they want to use
CLUSTERING <-function(A,K,method){
	if (method=='MCL_cs'){
		label <-MCL_cs(A)
		return(label)
	}else if (method =='SC_cs'){
		label <-SC_cs(A,K)
		return(label)
	}
}
```
To use MCL clustering, you may type
```{r call MCL}
labels1 <- CLUSTERING(A,method='MCL_cs')
```
where A is the adjacency matrix.
The output is cell labels:
```{r show labels}
head(labels1)
```
To use spectral clustering, you may type
```{r call spectral}
labels2 <- CLUSTERING(A,6,method='SC_cs')
```
where A is the adjacency matrix, '6' is the expected number of clusters.

If you have some reference labels, you can calculate the some external criteria to evaluate the agreement between predicted cell labels and reference labels. 

Note

* you may need to sort the predicted labels first
* make sure to that the cell names used in predicted label and reference labels are consistent 
```{r calculate ARI}
target <-read.table('C:/Users/Juan.XIe/Desktop/test1/Yan_cell.csv',header=T,sep=',')  # read reference label
head(target)
sorted <-labels1[match(target$Cell_type,names(labels1))] # sort the predicted label

ARI <-adjustedRandIndex(sorted,target$Cluster)  
RI <-adjustedRand(sorted,target$Cluster,randMethod='Rand')
FM <-adjustedRand(sorted,target$Cluster,randMethod='FM')
JI <-adjustedRand(sorted,target$Cluster,randMethod='Jaccard')
df <-data.frame(ARI=ARI, RandIndex=RI,FolkesMallow=FM, Jaccard=JI)
df
```
